File: matrix.js
class Matrix {
    constructor(mat) {
        this.matrix = mat
    }

    addRow(arr) {
        this.matrix.push(arr) //Para añadir una fila a nuestra matriz sólo hace falta añadir un array al final de ella
    }

    //Para añadir una columna hay que añadir al final de cada línea el valor de la columna en la posición de la fila
    addColumn(arr) {
        let rows = this.matrix.length

        for (let i = 0; i < rows; i++) {
            this.matrix[i].push(arr[i])
        }
    }

    //Para multiplicar una matriz por un número hay que multiplicar todos los elementos de la misma por ese número
    scalarMult(scalar) {
        let rows = this.matrix.length;
        let cols = this.matrix[0].length;

        for (let i = 0; i < rows; i++) {
            for (let j = 0; j < cols; j++) {
                this.matrix[i][j] *= scalar
            }
        }
    }

    //Para poder sumar dos matrices estas deben tener las mismas dimensiones. Hay que sumar cada numero con aquel que esté en la misma posición de la otra matriz
    sum(mat, op) {
        //Filas y columnas de nuestra matriz
        let rows = this.matrix.length;
        let cols = this.matrix[0].length;

        //Filas y columnas de la otra matriz
        let oRows = mat.length;
        let oCols = mat[0].length;

        //Se comprueba si la dimensión de las dos matrices es la misma
        if (rows != oRows || cols != oCols) {
            console.log('LAS MATRICES NO SON SUMABLES')
            return
        }

        //Si es suma, se suman
        if (op == '+') {
            for (let i = 0; i < rows; i++) {
                for (let j = 0; j < cols; j++) {
                    this.matrix[i][j] += mat[i][j]
                }
            }
        } else if (op == '-') {
            //Si es una resta se restan
            for (let i = 0; i < rows; i++) {
                for (let j = 0; j < cols; j++) {
                    this.matrix[i][j] -= mat[i][j]
                }
            }
        } else {
            //Si no es ni una suma ni una resta, se descarta la operación
            console.log('LA OPERACIÓN NO ESTÁ PERMITIDA')
        }
    }

    multiply(mat) {
        let rows = this.matrix.length; //Filas de esta matriz
        let cols = this.matrix[0].length; //Columnas de esta matriz
        let oRows = mat.length //Filas de la otra matriz
        let oCols = mat[0].length //Columnas de la otra matriz

        //Para que una matriz pueda ser multiplicada por otra el número de columnas de la primera debe ser igual al número de filas de la segunda
        if (cols != oRows) {
            console.log('LAS MATRICES NO PUEDEN SER MULTIPLICADAS')
            return
        }
        
        const result = new Array(rows); //Array donde se irán guardando los valores según se vayan calculando

        //Multiplicar cada numero (fila * columna) y sumarlos entre ellos -> Ponerlos en la posición del Array temporal en la posición en la que corresponda
        for (let i = 0; i < rows; i++) {
            result[i] = new Array(oCols);
            for (let j = 0; j < oCols; j++) {
                result[i][j] = 0;
                for (let k = 0; k < cols; k++) {
                    result[i][j] += this.matrix[i][k] * mat[k][j];
                }
            }
        }

        //Actualizar la matriz original con el resultado de la matriz nueva
        this.matrix = result
    }
}

let matriz = [
    [1, 2, 1],
    [2, 0, 1]
]
let matriz2 = [
    [1, 2],
    [2, 1],
    [0, 1]
]

const m = new Matrix(matriz)
m.multiply(matriz2)

console.log(m)

File: Matrix.java
/* Calculadora de matrices
 * Para definir la matriz principal Matrix matrizPrincipal = new Matrix(int[][])
 * Para sumar dos matrices matrizPrincipal.suma(<Matriz sumando>) (Matriz sumando = int[][])
 * Para restar es igual que en la suma pero con matrizPrincipal.resta(<Matriz sustraendo>)
 * Para multiplicar dos matrices matrixPrincipal.multiplicar(matrizB) (matrizB = int[][])
 * En el ejemplo aportado solo se definen dos matrices multiplicables
 */

public class Matrix {
    protected int[][] matrix;
    //Código principal que se ejecutará
    public static void main(String []args) {
        int[][] matriz = {
            {1,2,1},
            {2,0,1}
        };

        int[][] multiplicador = {
            {1,2},
            {2,1},
            {0,1}
        };

        Matrix m = new Matrix(matriz);
        m.multiplicar(multiplicador);
    }

    //Constructor de la class Matrix
    public Matrix(int[][] mat) {
        this.matrix = mat; //Guardar en la variable matrix la matriz introducida por el usuario
    }

    //Escribir la matriz en la consola, ya que por defecto se genera un código en lugar de la representación deseada
    public void WriteMatrix() {
        for(int i = 0; i<matrix.length; i++) {
            for(int j = 0; j<matrix[0].length; j++) {
                System.out.print(matrix[i][j]); //Escribir cada linea de la matriz
                System.out.print(',');
            }
            System.out.println(); //Al acabar la fila de la matriz hacer un salto de línea
        }
    }

    //Sumar dos matrices
    public void suma(int[][] sumando) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int oRows = sumando.length;
        int oCols = sumando[0].length;

        if(rows != oRows || cols != oCols) { return; } //Comprobar que las matrices sean de las mismas dimensiones, si no lo son no se pueden sumar
        

        for(int i = 0; i<rows; i++) {
            for(int j = 0; j<cols; j++) {
                matrix[i][j] += sumando[i][j];
            }
        }
        WriteMatrix();        
    }

    public void resta(int[][] sustraendo) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int oRows = sustraendo.length;
        int oCols = sustraendo[0].length;

        if(rows != oRows || cols != oCols) { return; } //Comprobar que las matrices sean de las mismas dimensiones, si no lo son no se pueden restar
        

        for(int i = 0; i<rows; i++) {
            for(int j = 0; j<cols; j++) {
                matrix[i][j] -= sustraendo[i][j];
            }
        }
        WriteMatrix();        
    }

    public void multiplicar(int[][] multiplicador) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int oRows = multiplicador.length;
        int oCols = multiplicador[0].length;

        if(cols != oRows) { return; } // Comprobar si el número de columnas de la matriz tiene el mismo número de filas que la segunda matriz, de modo que sean multiplicables

        int[][] result = new int[rows][oCols];

        for(int i = 0; i<rows; i++) {
            result[i] = new int[oCols];
            for(int j = 0; j<oCols; j++) {
                result[i][j] = 0;
                for(int k = 0; k<cols; k++) {
                    result[i][j] += matrix[i][k] * multiplicador[k][j];
                }
            }
        }

        matrix = result;

        WriteMatrix();
    }
}
